File: bloom_filter_test.f90
! gfortran -o bloom_filter_test bloom_filter_test.f90
!
! ./bloom_filter_test

program bloom_filter_test
    implicit none

    ! Parameters: m = bit array size, k = number of hashes
    integer, parameter :: m = 100
    integer, parameter :: k = 3
    logical :: bits(m)
    
    character(len=20) :: additions(6)
    character(len=20) :: tests(5)
    integer :: i, j, h_idx
    logical :: found

    ! Initialize bit array to false
    bits = .false.

    ! New Testament Items
    additions = [character(len=20) :: "Fisher-Net", "Alabaster-Jar", &
                 "Thorns-Crown", "Seamless-Robe", "Five-Loaves", "Two-Fish"]

    print *, "--- Fortran Bloom Filter: Seeding New Testament Items ---"
    
    ! Adding items
    do i = 1, 6
        print *, "Added: ", trim(additions(i))
        do j = 1, k
            h_idx = get_hash(trim(additions(i)), j, m)
            bits(h_idx) = .true.
        end do
    end do

    print *
    print *, "--- Testing Membership ---"
    
    tests = [character(len=20) :: "Fisher-Net", "Two-Fish", &
             "Golden-Calf", "Fisher-Boat", "Centurion-Spear"]

    do i = 1, 5
        found = .true.
        do j = 1, k
            h_idx = get_hash(trim(tests(i)), j, m)
            if (.not. bits(h_idx)) then
                found = .false.
                exit
            end if
        end do
        
        if (found) then
            print *, trim(tests(i)), ": MAYBE PRESENT"
        else
            print *, trim(tests(i)), ": DEFINITELY ABSENT"
        end if
    end do

contains

    ! Simple DJB2-style hash function
    integer function get_hash(str, seed, max_m)
        character(len=*), intent(in) :: str
        integer, intent(in) :: seed, max_m
        integer(8) :: h  ! Use 8-byte integer for intermediate math
        integer :: p

        h = 5381 + seed
        do p = 1, len(str)
            h = ((h * 33) + ichar(str(p:p)))
        end do
        
        ! Map to array index 1 to max_m
        get_hash = mod(abs(h), int(max_m, 8)) + 1
    end function get_hash

end program bloom_filter_test

File: BloomFilterTest.go
// go run BloomFilterTest.go

package main

import (
	"fmt"
	"hash/fnv"
)

// BloomFilterTest struct
type BloomFilterTest struct {
	bits      []byte
	m         uint32 // Total bits
	hashCount uint32 // Number of hashes (k)
}

// NewBloomFilterTest initializes a filter with m bits and k hashes
func NewBloomFilterTest(m, k uint32) *BloomFilterTest {
	// Size the slice to hold m bits (rounded up to nearest byte)
	return &BloomFilterTest{
		bits:      make([]byte, (m+7)/8),
		m:         m,
		hashCount: k,
	}
}

// getHash returns a bit index for a specific seed
func (bf *BloomFilterTest) getHash(item string, seed uint32) uint32 {
	h := fnv.New32a()
	h.Write([]byte(item))
	// We add the seed to the hash to simulate multiple hash functions
	return (h.Sum32() + seed*0x9e3779b9) % bf.m
}

// Add an item to the filter
func (bf *BloomFilterTest) Add(item string) {
	for i := uint32(0); i < bf.hashCount; i++ {
		idx := bf.getHash(item, i)
		// Locate the byte and the specific bit inside it
		bf.bits[idx/8] |= (1 << (idx % 8))
	}
}

// Exists checks if an item might be in the filter
func (bf *BloomFilterTest) Exists(item string) bool {
	for i := uint32(0); i < bf.hashCount; i++ {
		idx := bf.getHash(item, i)
		if bf.bits[idx/8]&(1<<(idx%8)) == 0 {
			return false // Definitely not there
		}
	}
	return true // Probably there
}

func main() {
	// 100 bits, 3 hashes
	bf := NewBloomFilterTest(100, 3)

	// New Testament Data
	additions := []string{"Fisher-Net", "Alabaster-Jar", "Thorns-Crown", "Seamless-Robe"}

	fmt.Println("--- Seeding Go Bloom Filter ---")
	for _, item := range additions {
		bf.Add(item)
		fmt.Printf("Added: %s\n", item)
	}

	fmt.Println("\n--- Testing Membership ---")
	tests := []string{"Fisher-Net", "Thorns-Crown", "Golden-Calf", "Centurion-Spear", "Fisher-Boat"}

	for _, test := range tests {
		status := "DEFINITELY ABSENT"
		if bf.Exists(test) {
			status = "PROBABLY PRESENT"
		}
		fmt.Printf("%-18s : %s\n", test, status)
	}
}
